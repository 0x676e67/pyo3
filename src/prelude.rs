// Copyright (c) 2017-present PyO3 Project and Contributors

//! A collection of items you most likely want to have in scope when working with pyo3
//!
//! The purpose of this module is to alleviate imports of many common pyo3 traits
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use conversion::{FromPyObject, IntoPyObject, PyTryFrom, PyTryInto, ToPyObject};
pub use err::{PyErr, PyResult};
pub use instance::{AsPyRef, Py, PyToken};
pub use noargs::NoArgs;
pub use object::PyObject;
pub use objectprotocol::ObjectProtocol;
pub use python::Python;
pub use pythonrun::GILGuard;
// This is only part of the prelude because we need it for the pymodinit function
pub use types::PyModule;
// This is required for the constructor
pub use PyRawObject;

pub use pyo3cls::{pyclass, pyfunction, pymethods, pyproto};

#[cfg(Py_3)]
pub use pyo3cls::mod3init as pymodinit;

#[cfg(not(Py_3))]
pub use pyo3cls::mod2init as pymodinit;
