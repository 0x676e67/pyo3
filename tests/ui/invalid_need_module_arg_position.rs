use pyo3::prelude::*;

#[pymodule]
fn module(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
    #[pyfn(m, pass_module)]
    fn fail<'py>(string: &str, module: &'py PyModule) -> PyResult<&'py str> {
        module.name()
    }
    Ok(())
}

fn main() {}
