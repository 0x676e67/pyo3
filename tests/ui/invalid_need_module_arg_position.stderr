error[E0277]: the trait bound `&str: From<&pyo3::prelude::PyModule>` is not satisfied
 --> tests/ui/invalid_need_module_arg_position.rs:6:26
  |
6 |     fn fail<'py>(string: &str, module: &'py PyModule) -> PyResult<&'py str> {
  |                          ^ the trait `From<&pyo3::prelude::PyModule>` is not implemented for `&str`
  |
  = help: the following other types implement trait `From<T>`:
            <String as From<char>>
            <String as From<Box<str>>>
            <String as From<Cow<'a, str>>>
            <String as From<&str>>
            <String as From<&mut str>>
            <String as From<&String>>
  = note: required for `&pyo3::prelude::PyModule` to implement `Into<&str>`
